src/compiler/binder.ts

```diff
@@ -278,6 +278,16 @@ namespace ts {
                     Debug.assert(isWellKnownSymbolSyntactically(nameExpression));
                     return getPropertyNameForKnownSymbolName(idText((<PropertyAccessExpression>nameExpression).name));
                 }
+                if (isPrivateName(name)) {
+                    // containingClass exists because private names only allowed inside classes
+                    const containingClass = getContainingClass(name.parent);

```

Comment:
Is there a way to determine this without using `.parent`? Then you wouldn't need to move the assignment of `.parent` up in `declareSymbol` (https://github.com/microsoft/TypeScript/pull/30829/files#r299688104). Normally the parent would just be `node`, except for the fact we also look for assigned names in `getNameOfDeclaration`, so you might need to do a test to ensure we're actually looking at the name of the node.

```ts
if (isPrivateName(name)) {
  if (isNamedDeclaration(node) && name === node.name) { // ensures this isn't an "assigned name"
    const containingClass = getContainingClass(node); // the parent of `name`
    ...
```

Reply:
Thanks for pointing this out.
- What is an assigned name? I only found one mention of this concept in the codebase and it wasn't enough for me to figure it out.

Asking so I can figure out what to do if `name` is an assigned name

Reply:
In `var f = function () { }`, the function expression has the assigned name `f`. getAssignedName is basically a helper for `getNameOfDeclaration`, but the other parts only look down the tree, not up.

It also applies to class expressions, but that's it.

Reply:
(outside of typescript this is called an "inferred" name, since it's not applied to all assignments and also not only applied to assignments)

Reply:
can you point me to the definition in the spec? "assigned name" is just an internal name *not* based on reading the spec :) I'd be interested in making Typescript's treatment closer to the spec treatment.

Reply:
The spec doesn't use any particular term for it; certain grammar productions call `SetFunctionName` with a specific parsed result.

Reply:
I took another look at this and it seems like there should be no issue. My concern was this case:
```ts
class A {
  #x = class {};
}
```
When binding the class expression, we will get the "assigned name" of `#x`, which doesn't belong to the class expression but rather belongs to `A`. I was concerned that `getContainingClass` might mistakenly return the class expression, however `getContainingClass` starts at the node's `.parent`, which would still result in the correct class being found.

As such, this issue can be considered closed.